File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        
        HashSet<String> type = new HashSet<>(Arrays.asList(gems));
        HashMap<String, Integer> gemCount = new HashMap<>();
        
        int answer[] = {0, gems.length};
        int left = 0;
        int right = 0;
        int max = type.size();
        
        while(right < gems.length){
            gemCount.put(gems[right], gemCount.getOrDefault(gems[right], 0) + 1);
            right++;
            
            while(gemCount.size() == max){
                // 기존 범위보다 작으면 갱신
                if(right - left < answer[1]- answer[0]){
                    answer[0] = left;
                    answer[1] = right;
                }
                // 범위보다 크면은 범위를 줄여보자
                if(gemCount.get(gems[left]) > 1){
                    gemCount.put(gems[left], gemCount.get(gems[left]) -1);
                }else{
                    gemCount.remove(gems[left]);
                }
                left++;
            }
            
        }
        
        answer[0] ++;
        
        return answer;
    }
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.js
function solution(gems) {
    
    const set = new Set(gems);
    const gemCount = new Map();
    
    const max = set.size;
    var answer = [0, gems.length];
    
    var left = 0;
    var right = 0;
    
    while(right < gems.length){
        gemCount.set(gems[right], (gemCount.get(gems[right]) || 0) + 1)
        right ++
        while(gemCount.size == max){
            // 구간이 더 짧으면 갱신
            if(right - left < answer[1] - answer[0]){
                answer[0] = left;
                answer[1] = right;
            }
            // 구간을 줄여보자
            if(gemCount.get(gems[left]) > 1){
                gemCount.set(gems[left], gemCount.get(gems[left]) -1)
            }else{
                gemCount.delete(gems[left]);
            }
            left++;
        }
    }
    
    answer[0]++;
    
    return answer;
}